### PR TITLE
libstd: support creation of anonymous pipe on WinXP/2K3

### PR DESCRIPTION
**Pull request by jsen-**
`PIPE_REJECT_REMOTE_CLIENTS` flag is not supported on Windows < VISTA, and every invocation of `anon_pipe` including attempts to pipe `std::process::Child`'s stdio fails.
This PR should work around this issue by performing a runtime check of windows version and conditionally omitting this flag on "XP and friends".

Getting the version should be probably moved out of the function `anon_pipe` itself (the OS version does not often change during runtime :) ), but:
 - I didn't find any precedent for this and assuming there's not much overhead (I hope windows does not perform any heuristics to find out it's own version, just fills couple of fields in the struct).
 - the code path is not especially performance sensitive anyway.

### CODE REVIEW

File: src/libstd/sys/windows/c.rs [hunk 1]
```diff
@@ -809,6 +810,16 @@ pub enum EXCEPTION_DISPOSITION {
     ExceptionCollidedUnwind
 }
 
+#[repr(C)]
+pub struct OSVERSIONINFO {
+  pub dwOSVersionInfoSize: DWORD,
```
Comment (alexcrichton) [inline src/libstd/sys/windows/c.rs]: Could this use 4-space tabs?

**Review by alexcrichton — COMMENTED** [review]
Looks good to me, thanks!

**Comment by xen0n** [comment]
Isn't this opening a security hole on these platforms, as the flag is presumably there for literally preventing remote connections to the pipes created? The correct behavior [seems doable on those legacy systems][so-link] though, albeit only with some ACL tricks, and I don't know if we really should go this further to support Tier-3 platforms like that.

[so-link]: http://stackoverflow.com/a/11033265

**Comment by xen0n** [comment]
Also, [the `GetVersionExW` function][msdn] is:

* neither heuristics-free (its behavior depends on existence of app manifest, at least),
* nor guaranteed to work under future releases of Windows,

so IMO this may require some degree of rethink. At least the version check could be unnecessary if we simply go ahead creating, checking `GetLastError` in case of failure and retrying with the compatible method if it's actually due to not supporting the flag.


[msdn]: https://msdn.microsoft.com/en-us/library/windows/desktop/ms724451(v=vs.85).aspx

**Comment by jsen-** [comment]
@xen0n you're right, I missassumed the possible performance implication of `GetVersionExW` it should be at least moved outside of the function, but the truth is, retrying on `Parameter is incorrect` error should cover this better - this is easy

I can even try to hack something in the spirit of the linked SO answer later today, but I'm by no means a winapi expert

@alexcrichton sorry, didn't notice - rustfmt stopped working for me a while back and I haven't attempted to resurrect it yet, will fix in the next commit

**Comment by alexcrichton** [comment]
@xen0n's solution also sounds fine by me (try the current strategy and fall back if that fails). I'd be fine punting on ACL business and whatnot, XP is only Tier 3 after all

**Comment by xen0n** [comment]
@alexcrichton Sure, XP is Tier 3 and ACL is a lot of unneeded complexity, but IMO the security implication should at least be documented. People are bound to leverage the compatibility if it's there, and (incorrectly) expect the security level to be the same as the Tier-1 Windows versions. Although XP users' problems are much deeper than some under-protected pipes could bring about, I certainly don't think Rust should take the blame in case the introduced inconsistent behavior actually leads to exploits.

**Comment by sfackler** [comment]
Windows XP hasn't received security updates for over 2 years.

**Comment by xen0n** [comment]
@sfackler Yeah, sure. XP's overall security won't get any better even if we implemented the correct behavior, and since it's unsupported not documenting the inconsistency is perfectly acceptable as well. I'd then suggest putting a comment about the decision and forget about it. 😂

**Comment by jsen-** [comment]
Alright, I made a rough draft of how it might look like...
Tests on `i686-pc-windows-gnu` passed fine on my machine, but I have no idea how or whether it is even possible to somehow run the tests on WinXP (except that my code using it still works there after this change).

!!!**Currently the code leaks memory**!!! so it's not ready for prime time, but I'd like to ask for help from someone more experienced with this winapi stuff.
My basic question is when is it safe to free the pointers allocated by `AllocateAndInitializeSid` and `LocalAlloc` (is it safe to free right after the pipe was created or after not being used anymore?).

I believe the code is not very idiomatic rust, so all comments are very welcome.
Most of the attributes on FFI imports were guesswork, so it definitely needs attention as well.

**Comment by alexcrichton** [comment]
@jsen- I think @sfackler's got a [good point](https://github.com/rust-lang/rust/pull/37677#issuecomment-259733338) that guaranteeing the same semantics here is probably the last of one's "security concerns" on XP. In that sense what do you think of the strategy of attempting to pass the flag, and then trying to not pass the flag if that fails with a particular error? (for XP)

**Comment by jsen-** [comment]
@alexcrichton fully agree, I think this PR got a little overkill for what the issue deserves. Let me shake of some of it's weight
I'll try to push an update when I'm back from work

**Comment by jsen-** [comment]
Here's the original simple version minus windows version check. Relying on `CreateNamedPipeW` with `PIPE_REJECT_REMOTE_CLIENTS` returning `ERROR_INVALID_PARAMETER` on XP as suggested by @xen0n.
Note added to the documentation.

**Comment by alexcrichton** [comment]
Thanks! Looks good to me, but I think we can remove the note for XP users. We typically don't have this sort of platform-specific notes in the docs just yet, and this is a bit prominent for something that's relevant to very few people.

**Comment by xen0n** [comment]
Putting the note (preferably along with justification of security reduction) in code comment should be enough IMO.

**Comment by jsen-** [comment]
Removed the note from documentation and added code comment with implications and justification.
Sorry for the number of commits, I tried squashing some, but it seems to have just added more and more (poor git skills indeed)

**Comment by xen0n** [comment]
@jsen- Don't worry and just use `git rebase -i <your upstream remote>/master` for the nice fixups!

**Comment by alexcrichton** [comment]
Ah yeah code looks good to me, and with a squash I'd r+. Let me know if you need help squashing though

**Comment by jsen-** [comment]
okay I think it's good now, thanks for patience :)

**Pull request merged**